File: csharp/Maze.cs
using System;
using System.Collections.Generic;

public class MazeGenerator : Benchmark
{
    public enum CellKind
    {
        Wall = 0,
        Space = 1,
        Start = 2,
        Finish = 3,
        Border = 4,
        Path = 5
    }

    public class Cell
    {
        public CellKind Kind;
        public List<Cell> Neighbors;
        public int X;
        public int Y;

        public Cell(int x, int y)
        {
            Kind = CellKind.Wall;
            X = x;
            Y = y;
            Neighbors = new List<Cell>(4);
        }

        public bool IsWalkable() =>
            Kind == CellKind.Space || Kind == CellKind.Start || Kind == CellKind.Finish;

        public void Reset()
        {
            if (Kind == CellKind.Space)
                Kind = CellKind.Wall;
        }
    }

    public class Maze
    {
        private readonly int _width;
        private readonly int _height;
        private readonly Cell[,] _cells;
        private readonly Cell _start;
        private readonly Cell _finish;
        private readonly Random _random = new Random();

        public Maze(int width, int height)
        {
            _width = Math.Max(width, 5);
            _height = Math.Max(height, 5);
            _cells = new Cell[_height, _width];

            for (int y = 0; y < _height; y++)
                for (int x = 0; x < _width; x++)
                    _cells[y, x] = new Cell(x, y);

            _start = _cells[1, 1];
            _finish = _cells[_height - 2, _width - 2];
            _start.Kind = CellKind.Start;
            _finish.Kind = CellKind.Finish;

            UpdateNeighbors();
        }

        public void UpdateNeighbors()
        {
            for (int y = 0; y < _height; y++)
            {
                for (int x = 0; x < _width; x++)
                {
                    var cell = _cells[y, x];
                    cell.Neighbors.Clear();

                    if (x > 0 && y > 0 && x < _width - 1 && y < _height - 1)
                    {
                        cell.Neighbors.Add(_cells[y - 1, x]);
                        cell.Neighbors.Add(_cells[y + 1, x]);
                        cell.Neighbors.Add(_cells[y, x + 1]);
                        cell.Neighbors.Add(_cells[y, x - 1]);

                        for (int t = 0; t < 4; t++)
                        {
                            int i = Helper.NextInt(4);
                            int j = Helper.NextInt(4);
                            if (i != j)
                            {
                                var temp = cell.Neighbors[i];
                                cell.Neighbors[i] = cell.Neighbors[j];
                                cell.Neighbors[j] = temp;
                            }
                        }
                    }
                    else
                    {
                        cell.Kind = CellKind.Border;
                    }
                }
            }
        }

        public void Reset()
        {
            foreach (var cell in _cells)
                cell.Reset();

            _start.Kind = CellKind.Start;
            _finish.Kind = CellKind.Finish;
        }

        private void Dig(Cell startCell)
        {
            var stack = new Stack<Cell>();
            stack.Push(startCell);

            while (stack.Count > 0)
            {
                var cell = stack.Pop();

                int walkable = 0;
                foreach (var n in cell.Neighbors)
                    if (n.IsWalkable()) walkable++;

                if (walkable != 1) continue;

                cell.Kind = CellKind.Space;

                foreach (var n in cell.Neighbors)
                    if (n.Kind == CellKind.Wall)
                        stack.Push(n);
            }
        }

        private void EnsureOpenFinish(Cell startCell)
        {
            var stack = new Stack<Cell>();
            stack.Push(startCell);

            while (stack.Count > 0)
            {
                var cell = stack.Pop();

                cell.Kind = CellKind.Space;

                int walkable = 0;
                foreach (var n in cell.Neighbors)
                    if (n.IsWalkable()) walkable++;

                if (walkable > 1) continue;

                foreach (var n in cell.Neighbors)
                    if (n.Kind == CellKind.Wall)
                        stack.Push(n);
            }
        }

        public void Generate()
        {
            foreach (var n in _start.Neighbors)
                if (n.Kind == CellKind.Wall)
                    Dig(n);

            foreach (var n in _finish.Neighbors)
                if (n.Kind == CellKind.Wall)
                    EnsureOpenFinish(n);
        }

        public Cell GetStart() => _start;
        public Cell GetFinish() => _finish;
        public Cell MiddleCell() => _cells[_height / 2, _width / 2];

        public uint Checksum()
        {
            uint hasher = 2166136261;
            uint prime = 16777619;

            for (int y = 0; y < _height; y++)
                for (int x = 0; x < _width; x++)
                    if (_cells[y, x].Kind == CellKind.Space)
                    {
                        uint val = (uint)(x * y);
                        hasher = (hasher ^ val) * prime;
                    }

            return hasher;
        }

        public void PrintToConsole()
        {
            for (int y = 0; y < _height; y++)
            {
                for (int x = 0; x < _width; x++)
                {
                    switch (_cells[y, x].Kind)
                    {
                        case CellKind.Space: Console.Write(" "); break;
                        case CellKind.Wall: Console.Write("\u001b[34m#\u001b[0m"); break;
                        case CellKind.Border: Console.Write("\u001b[31mO\u001b[0m"); break;
                        case CellKind.Start: Console.Write("\u001b[32m>\u001b[0m"); break;
                        case CellKind.Finish: Console.Write("\u001b[32m<\u001b[0m"); break;
                        case CellKind.Path: Console.Write("\u001b[33m.\u001b[0m"); break;
                    }
                }
                Console.WriteLine();
            }
            Console.WriteLine();
        }

        public Cell GetCell(int x, int y)
        {
            if (x >= 0 && x < _width && y >= 0 && y < _height)
                return _cells[y, x];
            return null;
        }
    }

    private uint _result;
    private readonly int _width;
    private readonly int _height;
    private readonly Maze _maze;

    public MazeGenerator()
    {
        _width = (int)ConfigVal("w");
        _height = (int)ConfigVal("h");
        _maze = new Maze(_width, _height);
        _result = 0;
    }

    public override void Prepare() { }

    public override void Run(long iterationId)
    {
        _maze.Reset();
        _maze.Generate();
        _result += (uint)_maze.MiddleCell().Kind;
    }

    public override uint Checksum => _result + _maze.Checksum();
    public override string TypeName => "Maze::Generator";
}

public class MazeBFS : Benchmark
{
    private uint _result;
    private readonly int _width;
    private readonly int _height;
    private readonly MazeGenerator.Maze _maze;
    private List<MazeGenerator.Cell> _path = new List<MazeGenerator.Cell>();

    public MazeBFS()
    {
        _width = (int)ConfigVal("w");
        _height = (int)ConfigVal("h");
        _maze = new MazeGenerator.Maze(_width, _height);
        _result = 0;
    }

    public override void Prepare()
    {
        _maze.Generate();
    }

    private List<MazeGenerator.Cell> Bfs(MazeGenerator.Cell start, MazeGenerator.Cell target)
    {
        if (start == target)
            return new List<MazeGenerator.Cell> { start };

        var queue = new Queue<int>();
        var visited = new bool[_height, _width];
        var path = new List<(MazeGenerator.Cell cell, int parent)>();

        visited[start.Y, start.X] = true;
        path.Add((start, -1));
        queue.Enqueue(0);

        while (queue.Count > 0)
        {
            int pathId = queue.Dequeue();
            var cell = path[pathId].cell;

            foreach (var neighbor in cell.Neighbors)
            {
                if (neighbor == target)
                {
                    var result = new List<MazeGenerator.Cell> { target };
                    int current = pathId;
                    while (current >= 0)
                    {
                        result.Add(path[current].cell);
                        current = path[current].parent;
                    }
                    result.Reverse();
                    return result;
                }

                if (neighbor.IsWalkable() && !visited[neighbor.Y, neighbor.X])
                {
                    visited[neighbor.Y, neighbor.X] = true;
                    path.Add((neighbor, pathId));
                    queue.Enqueue(path.Count - 1);
                }
            }
        }

        return new List<MazeGenerator.Cell>();
    }

    private uint MidCellChecksum(List<MazeGenerator.Cell> p)
    {
        if (p.Count == 0) return 0;
        var cell = p[p.Count / 2];
        return (uint)(cell.X * cell.Y);
    }

    public override void Run(long iterationId)
    {
        _path = Bfs(_maze.GetStart(), _maze.GetFinish());
        _result += (uint)_path.Count;
    }

    public override uint Checksum => _result + MidCellChecksum(_path);
    public override string TypeName => "Maze::BFS";
}

public class MazeAStar : Benchmark
{
    private uint _result;
    private readonly int _width;
    private readonly int _height;
    private readonly MazeGenerator.Maze _maze;
    private List<MazeGenerator.Cell> _path = new List<MazeGenerator.Cell>();

    public MazeAStar()
    {
        _width = (int)ConfigVal("w");
        _height = (int)ConfigVal("h");
        _maze = new MazeGenerator.Maze(_width, _height);
        _result = 0;
    }

    public override void Prepare()
    {
        _maze.Generate();
    }

    private int Heuristic(MazeGenerator.Cell a, MazeGenerator.Cell b) =>
        Math.Abs(a.X - b.X) + Math.Abs(a.Y - b.Y);

    private int Idx(int y, int x) => y * _width + x;

    private List<MazeGenerator.Cell> AStar(MazeGenerator.Cell start, MazeGenerator.Cell target)
    {
        if (start == target)
            return new List<MazeGenerator.Cell> { start };

        int size = _width * _height;

        var cameFrom = new int[size];
        var gScore = new int[size];
        var bestF = new int[size];

        for (int i = 0; i < size; i++)
        {
            cameFrom[i] = -1;
            gScore[i] = int.MaxValue;
            bestF[i] = int.MaxValue;
        }

        int startIdx = Idx(start.Y, start.X);
        int targetIdx = Idx(target.Y, target.X);

        var openSet = new PriorityQueue<int, int>();

        gScore[startIdx] = 0;
        int fStart = Heuristic(start, target);
        openSet.Enqueue(startIdx, fStart);
        bestF[startIdx] = fStart;

        while (openSet.Count > 0)
        {
            int currentIdx = openSet.Dequeue();

            if (gScore[currentIdx] == int.MaxValue) continue;

            if (currentIdx == targetIdx)
            {
                var result = new List<MazeGenerator.Cell>();
                int cur = currentIdx;
                while (cur != -1)
                {
                    int y = cur / _width;
                    int x = cur % _width;
                    result.Add(_maze.GetCell(x, y));
                    cur = cameFrom[cur];
                }
                result.Reverse();
                return result;
            }

            int currentY = currentIdx / _width;
            int currentX = currentIdx % _width;
            var current = _maze.GetCell(currentX, currentY);
            int currentG = gScore[currentIdx];

            foreach (var neighbor in current.Neighbors)
            {
                if (!neighbor.IsWalkable()) continue;

                int neighborIdx = Idx(neighbor.Y, neighbor.X);
                int tentativeG = currentG + 1;

                if (tentativeG < gScore[neighborIdx])
                {
                    cameFrom[neighborIdx] = currentIdx;
                    gScore[neighborIdx] = tentativeG;
                    int fNew = tentativeG + Heuristic(neighbor, target);

                    if (fNew < bestF[neighborIdx])
                    {
                        bestF[neighborIdx] = fNew;
                        openSet.Enqueue(neighborIdx, fNew);
                    }
                }
            }
        }

        return new List<MazeGenerator.Cell>();
    }

    private uint MidCellChecksum(List<MazeGenerator.Cell> p)
    {
        if (p.Count == 0) return 0;
        var cell = p[p.Count / 2];
        return (uint)(cell.X * cell.Y);
    }

    public override void Run(long iterationId)
    {
        _path = AStar(_maze.GetStart(), _maze.GetFinish());
        _result += (uint)_path.Count;
    }

    public override uint Checksum => _result + MidCellChecksum(_path);
    public override string TypeName => "Maze::AStar";
}

File: fsharp/Maze.fs
namespace Benchmarks

open System
open System.Collections.Generic

module MazeTypes =
    type CellKind =
        | Wall = 0
        | Space = 1
        | Start = 2
        | Finish = 3
        | Border = 4
        | Path = 5

    type Cell(x: int, y: int) =
        let mutable kind = CellKind.Wall
        let neighbors = ResizeArray<Cell>(4)  

        member _.X = x
        member _.Y = y
        member _.Kind with get() = kind and set(value) = kind <- value
        member _.Neighbors = neighbors

        member this.AddNeighbor(cell: Cell) =
            neighbors.Add(cell)

        member this.IsWalkable() =
            kind = CellKind.Space || kind = CellKind.Start || kind = CellKind.Finish

        member this.Reset() =
            if kind = CellKind.Space then
                kind <- CellKind.Wall

    type Maze(width: int, height: int) as this =
        let w = max width 5
        let h = max height 5
        let cells = Array2D.init h w (fun y x -> Cell(x, y))
        let start = cells.[1, 1]
        let finish = cells.[h-2, w-2]

        do
            start.Kind <- CellKind.Start
            finish.Kind <- CellKind.Finish
            this.UpdateNeighbors()

        member _.Width = w
        member _.Height = h
        member _.Cells = cells
        member _.Start = start
        member _.Finish = finish

        member this.UpdateNeighbors() =
            for y = 0 to h - 1 do
                for x = 0 to w - 1 do
                    cells.[y, x].Neighbors.Clear()

            for y = 0 to h - 1 do
                for x = 0 to w - 1 do
                    let cell = cells.[y, x]

                    if x > 0 && y > 0 && x < w - 1 && y < h - 1 then
                        cell.AddNeighbor(cells.[y-1, x])
                        cell.AddNeighbor(cells.[y+1, x])
                        cell.AddNeighbor(cells.[y, x+1])
                        cell.AddNeighbor(cells.[y, x-1])

                        for _ = 1 to 4 do
                            let i = Helper.NextInt(4)
                            let j = Helper.NextInt(4)
                            if i <> j then
                                let temp = cell.Neighbors.[i]
                                cell.Neighbors.[i] <- cell.Neighbors.[j]
                                cell.Neighbors.[j] <- temp
                    else
                        cell.Kind <- CellKind.Border

        member this.Reset() =
            for y = 0 to h - 1 do
                for x = 0 to w - 1 do
                    cells.[y, x].Reset()
            start.Kind <- CellKind.Start
            finish.Kind <- CellKind.Finish

        member this.Dig(startCell: Cell) =
            let stack = Stack<Cell>()
            stack.Push(startCell)

            while stack.Count > 0 do
                let cell = stack.Pop()

                let mutable walkable = 0
                for i = 0 to cell.Neighbors.Count - 1 do
                    if cell.Neighbors.[i].IsWalkable() then
                        walkable <- walkable + 1

                if walkable = 1 then
                    cell.Kind <- CellKind.Space

                    for i = 0 to cell.Neighbors.Count - 1 do
                        if cell.Neighbors.[i].Kind = CellKind.Wall then
                            stack.Push(cell.Neighbors.[i])

        member this.EnsureOpenFinish(startCell: Cell) =
            let stack = Stack<Cell>()
            stack.Push(startCell)

            while stack.Count > 0 do
                let cell = stack.Pop()

                cell.Kind <- CellKind.Space

                let mutable walkable = 0
                for i = 0 to cell.Neighbors.Count - 1 do
                    if cell.Neighbors.[i].IsWalkable() then
                        walkable <- walkable + 1

                if walkable <= 1 then
                    for i = 0 to cell.Neighbors.Count - 1 do
                        if cell.Neighbors.[i].Kind = CellKind.Wall then
                            stack.Push(cell.Neighbors.[i])

        member this.Generate() =
            for n in start.Neighbors do
                if n.Kind = CellKind.Wall then
                    this.Dig(n)

            for n in finish.Neighbors do
                if n.Kind = CellKind.Wall then
                    this.EnsureOpenFinish(n)

        member this.MiddleCell() =
            cells.[h / 2, w / 2]

        member this.Checksum() =
            let mutable hasher = 2166136261u
            let prime = 16777619u

            for y = 0 to h - 1 do
                for x = 0 to w - 1 do
                    let cell = cells.[y, x]
                    if cell.Kind = CellKind.Space then
                        let val' = uint32 (x * y)
                        hasher <- (hasher ^^^ val') * prime
            hasher

        member this.PrintToConsole() =
            for y = 0 to h - 1 do
                for x = 0 to w - 1 do
                    match cells.[y, x].Kind with
                    | CellKind.Space -> printf " "
                    | CellKind.Wall -> printf "\u001B[34m#\u001B[0m"
                    | CellKind.Border -> printf "\u001B[31mO\u001B[0m"
                    | CellKind.Start -> printf "\u001B[32m>\u001B[0m"
                    | CellKind.Finish -> printf "\u001B[32m<\u001B[0m"
                    | CellKind.Path -> printf "\u001B[33m.\u001B[0m"
                    | _ -> printf "?"
                printfn ""
            printfn ""

type BfsPathNode(cell: MazeTypes.Cell, parent: int) =
    member _.Cell = cell
    member _.Parent = parent

type MazeGenerator() =
    inherit Benchmark()

    let mutable width = 0
    let mutable height = 0
    let mutable maze : MazeTypes.Maze option = None
    let mutable resultVal = 0u

    override this.Prepare() =
        width <- int (this.ConfigVal("w"))
        height <- int (this.ConfigVal("h"))
        let newMaze = MazeTypes.Maze(width, height)
        maze <- Some newMaze
        resultVal <- 0u

    override this.Run(_: int64) =
        match maze with
        | Some m ->
            m.Reset()
            m.Generate()
            resultVal <- resultVal + uint32 (m.MiddleCell().Kind)
        | None -> ()

    override this.Checksum =
        match maze with
        | Some m -> resultVal + m.Checksum()
        | None -> 0u
    override this.Name = "Maze::Generator"

type MazeBFS() =
    inherit Benchmark()

    let mutable width = 0
    let mutable height = 0
    let mutable maze : MazeTypes.Maze option = None
    let mutable resultVal = 0u
    let mutable path : MazeTypes.Cell list = []

    override this.Prepare() =
        width <- int (this.ConfigVal("w"))
        height <- int (this.ConfigVal("h"))
        let newMaze = MazeTypes.Maze(width, height)
        newMaze.Generate()
        maze <- Some newMaze
        resultVal <- 0u
        path <- []

    member this.Bfs(start: MazeTypes.Cell, target: MazeTypes.Cell) =
        if start = target then [start] else

        let queue = Queue<int>()
        let visited = Array2D.zeroCreate<bool> height width
        let pathNodes = ResizeArray<BfsPathNode>()

        visited.[start.Y, start.X] <- true
        pathNodes.Add(BfsPathNode(start, -1))
        queue.Enqueue(0)

        let mutable result = []

        while queue.Count > 0 && result.IsEmpty do
            let pathId = queue.Dequeue()
            let node = pathNodes.[pathId]

            for i = 0 to node.Cell.Neighbors.Count - 1 do
                let neighbor = node.Cell.Neighbors.[i]
                if neighbor = target then
                    let mutable cur = pathId
                    let mutable res = [target]
                    while cur >= 0 do
                        res <- pathNodes.[cur].Cell :: res
                        cur <- pathNodes.[cur].Parent
                    result <- res |> List.rev
                else
                    if neighbor.IsWalkable() && not visited.[neighbor.Y, neighbor.X] then
                        visited.[neighbor.Y, neighbor.X] <- true
                        pathNodes.Add(BfsPathNode(neighbor, pathId))
                        queue.Enqueue(pathNodes.Count - 1)

        result

    member this.MidCellChecksum(path: MazeTypes.Cell list) =
        if path.IsEmpty then 0u
        else
            let cell = path.[path.Length / 2]
            uint32 (cell.X * cell.Y)

    override this.Run(_: int64) =
        match maze with
        | Some m ->
            path <- this.Bfs(m.Start, m.Finish)
            resultVal <- resultVal + uint32 path.Length
        | None -> ()

    override this.Checksum = resultVal + this.MidCellChecksum(path)
    override this.Name = "Maze::BFS"

type MazeAStar() =
    inherit Benchmark()

    let mutable width = 0
    let mutable height = 0
    let mutable maze : MazeTypes.Maze option = None
    let mutable resultVal = 0u
    let mutable path : MazeTypes.Cell list = []

    override this.Prepare() =
        width <- int (this.ConfigVal("w"))
        height <- int (this.ConfigVal("h"))
        let newMaze = MazeTypes.Maze(width, height)
        newMaze.Generate()
        maze <- Some newMaze
        resultVal <- 0u
        path <- []

    member this.Heuristic(a: MazeTypes.Cell, b: MazeTypes.Cell) =
        abs (a.X - b.X) + abs (a.Y - b.Y)

    member this.Idx(y: int, x: int) = y * width + x

    member this.AStar(start: MazeTypes.Cell, target: MazeTypes.Cell) =
        if start = target then [start] else

        let size = width * height
        let cameFrom = Array.create size -1
        let gScore = Array.create size Int32.MaxValue
        let bestF = Array.create size Int32.MaxValue

        let startIdx = this.Idx(start.Y, start.X)
        let targetIdx = this.Idx(target.Y, target.X)

        let openSet = PriorityQueue<int, int>()
        let inOpen = Array.create size false

        let fStart = this.Heuristic(start, target)
        openSet.Enqueue(startIdx, fStart)
        bestF.[startIdx] <- fStart
        gScore.[startIdx] <- 0
        inOpen.[startIdx] <- true

        let mutable result = []

        while openSet.Count > 0 && result.IsEmpty do
            let ok, currentIdx, _ = openSet.TryDequeue()  
            if ok then
                inOpen.[currentIdx] <- false

                if currentIdx = targetIdx then
                    let mutable cur = currentIdx
                    while cur <> -1 do
                        let y = cur / width
                        let x = cur % width
                        match maze with
                        | Some m -> result <- m.Cells.[y, x] :: result
                        | None -> ()
                        cur <- cameFrom.[cur]
                    result <- result |> List.rev
                else
                    let currentY = currentIdx / width
                    let currentX = currentIdx % width
                    match maze with
                    | Some m ->
                        let currentCell = m.Cells.[currentY, currentX]
                        let currentG = gScore.[currentIdx]

                        for i = 0 to currentCell.Neighbors.Count - 1 do
                            let neighbor = currentCell.Neighbors.[i]
                            if neighbor.IsWalkable() then
                                let neighborIdx = this.Idx(neighbor.Y, neighbor.X)
                                let tentativeG = currentG + 1

                                if tentativeG < gScore.[neighborIdx] then
                                    cameFrom.[neighborIdx] <- currentIdx
                                    gScore.[neighborIdx] <- tentativeG
                                    let fNew = tentativeG + this.Heuristic(neighbor, target)

                                    if fNew < bestF.[neighborIdx] then
                                        bestF.[neighborIdx] <- fNew
                                        openSet.Enqueue(neighborIdx, fNew)
                                        inOpen.[neighborIdx] <- true
                    | None -> ()

        result

    member this.MidCellChecksum(path: MazeTypes.Cell list) =
        if path.IsEmpty then 0u
        else
            let cell = path.[path.Length / 2]
            uint32 (cell.X * cell.Y)

    override this.Run(_: int64) =
        match maze with
        | Some m ->
            path <- this.AStar(m.Start, m.Finish)
            resultVal <- resultVal + uint32 path.Length
        | None -> ()

    override this.Checksum = resultVal + this.MidCellChecksum(path)
    override this.Name = "Maze::AStar"